presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java

```diff
@@ -800,4 +800,34 @@ public void testWidthBucketArray()
         // this is a case that we can't catch because we are using binary search to bisect the bins array
         assertFunction("width_bucket(1.5, array[1.0, 2.3, 2.0])", BIGINT, 1L);
     }
+
+    @Test
+    public void testCosineSimilarity()
+    {
+        assertFunction("cosine_similarity( map(array ['a', 'b'], array [1.0, 2.0]), map(array ['c', 'b'], array [1.0, 3.0]) )",
+                DOUBLE,
+                2 * 3 / (Math.sqrt(5) * Math.sqrt(10)));
+
+        assertFunction("cosine_similarity( map(array ['a', 'b', 'c'], array [1.0, 2.0, -1.0]), map(array ['c', 'b'], array [1.0, 3.0]) )",
+                DOUBLE,
+                (2 * 3 + (-1) * 1) / (Math.sqrt(1 + 4 + 1) * Math.sqrt(1 + 9)));
+
+        assertFunction("cosine_similarity( map(array ['a', 'b', 'c'], array [1.0, 2.0, -1.0]), map(array ['d', 'e'], array [1.0, 3.0]) )",
+                DOUBLE,
+                0.0);
+
+        assertFunction("cosine_similarity( null, map(array ['c', 'b'], array [1.0, 3.0]) )",
+                DOUBLE,
+                null);
+
+        assertFunction("cosine_similarity( map(array ['a', 'b'], array [1.0, null]), map(array ['c', 'b'], array [1.0, 3.0]) )",
+                DOUBLE,
+                null);
+
+        assertInvalidFunction("cosine_similarity( map(array [null, 'b'], array [1.0, 2.4]), map(array ['c', 'b'], array [1.0, 3.0]) )",
```

Comment:
@cberner this error message doesn't come from the UDF. I don't think we should test for it.